File: pkg/sentry/control/logging.go
// Copyright 2019 The gVisor Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package control

import (
	"fmt"

	"gvisor.dev/gvisor/pkg/log"
	"gvisor.dev/gvisor/pkg/sentry/strace"
	"gvisor.dev/gvisor/pkg/tcpip/link/sniffer"
)

// LoggingArgs are the arguments to use for changing the logging
// level and strace list.
type LoggingArgs struct {
	// SetLevel is a flag used to indicate that we should update
	// the logging level. We should be able to change the strace
	// list without affecting the logging level and vice versa.
	SetLevel bool

	// Level is the log level that will be set if SetLevel is true.
	Level log.Level

	// SetLogPackets indicates that we should update the log packets flag.
	SetLogPackets bool

	// LogPackets is the actual value to set for LogPackets.
	// SetLogPackets must be enabled to indicate that we're changing
	// the value.
	LogPackets bool

	// SetStrace is a flag used to indicate that strace related
	// arguments were passed in.
	SetStrace bool

	// EnableStrace is a flag from the CLI that specifies whether to
	// enable strace at all. If this flag is false then a completely
	// pristine copy of the syscall table will be swapped in. This
	// approach is used to remain consistent with an empty strace
	// allowlist meaning trace all system calls.
	EnableStrace bool

	// Strace is the allowlist of syscalls to trace to log. If this
	// and StraceEventAllowlist are empty trace all system calls.
	StraceAllowlist []string

	// SetEventStrace is a flag used to indicate that event strace
	// related arguments were passed in.
	SetEventStrace bool

	// StraceEventAllowlist is the allowlist of syscalls to trace
	// to event log.
	StraceEventAllowlist []string
}

// Logging provides functions related to logging.
type Logging struct{}

// Change will change the log level and strace arguments. Although
// this functions signature requires an error it never actually
// returns an error. It's required by the URPC interface.
// Additionally, it may look odd that this is the only method
// attached to an empty struct but this is also part of how
// URPC dispatches.
func (l *Logging) Change(args *LoggingArgs, code *int) error {
	if args.SetLevel {
		// Logging uses an atomic for the level so this is thread safe.
		log.SetLevel(args.Level)
	}

	if args.SetLogPackets {
		if args.LogPackets {
			sniffer.LogPackets.Store(1)
		} else {
			sniffer.LogPackets.Store(0)
		}
		log.Infof("LogPackets set to: %v", sniffer.LogPackets.Load())
	}

	if args.SetStrace {
		if err := l.configureStrace(args); err != nil {
			return fmt.Errorf("error configuring strace: %v", err)
		}
	}

	if args.SetEventStrace {
		if err := l.configureEventStrace(args); err != nil {
			return fmt.Errorf("error configuring event strace: %v", err)
		}
	}

	return nil
}

func (l *Logging) configureStrace(args *LoggingArgs) error {
	if args.EnableStrace {
		// Install the allowlist specified.
		if len(args.StraceAllowlist) > 0 {
			if err := strace.Enable(args.StraceAllowlist, strace.SinkTypeLog); err != nil {
				return err
			}
		} else {
			// For convenience, if strace is enabled but allowlist
			// is empty, enable everything to log.
			strace.EnableAll(strace.SinkTypeLog)
		}
	} else {
		// Uninstall all strace functions.
		strace.Disable(strace.SinkTypeLog)
	}
	return nil
}

func (l *Logging) configureEventStrace(args *LoggingArgs) error {
	if len(args.StraceEventAllowlist) > 0 {
		if err := strace.Enable(args.StraceEventAllowlist, strace.SinkTypeEvent); err != nil {
			return err
		}
	} else {
		strace.Disable(strace.SinkTypeEvent)
	}
	return nil
}

File: test/util/logging.cc
// Copyright 2018 The gVisor Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "test/util/logging.h"

#include <errno.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>

namespace gvisor {
namespace testing {

namespace {

// We implement this here instead of using test_util to avoid cyclic
// dependencies.
int Write(int fd, const char* buf, size_t size) {
  size_t written = 0;
  while (written < size) {
    int res = write(fd, buf + written, size - written);
    if (res < 0 && errno == EINTR) {
      continue;
    } else if (res <= 0) {
      break;
    }

    written += res;
  }
  return static_cast<int>(written);
}

// Write 32-bit decimal number to fd.
int WriteNumber(int fd, uint32_t val) {
  constexpr char kDigits[] = "0123456789";
  constexpr int kBase = 10;

  // 10 chars for 32-bit number in decimal, 1 char for the NUL-terminator.
  constexpr int kBufferSize = 11;
  char buf[kBufferSize];

  // Convert the number to string.
  char* s = buf + sizeof(buf) - 1;
  size_t size = 0;

  *s = '\0';
  do {
    s--;
    size++;

    *s = kDigits[val % kBase];
    val /= kBase;
  } while (val);

  return Write(fd, s, size);
}

}  // namespace

void CheckFailure(const char* cond, size_t cond_size, const char* msg,
                  size_t msg_size, int errno_value) {
  constexpr char kCheckFailure[] = "Check failed: ";
  Write(2, kCheckFailure, sizeof(kCheckFailure) - 1);
  Write(2, cond, cond_size);

  if (msg != nullptr) {
    Write(2, ": ", 2);
    Write(2, msg, msg_size);
  }

  if (errno_value != 0) {
    constexpr char kErrnoMessage[] = " (errno ";
    Write(2, kErrnoMessage, sizeof(kErrnoMessage) - 1);
    WriteNumber(2, errno_value);
    Write(2, ")", 1);
  }

  Write(2, "\n", 1);

  abort();
}

}  // namespace testing
}  // namespace gvisor
